### PR TITLE
Http updates

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
- [ ] Upgrade guidance
- [x] changelog

To allow seamless adoption of the incoming `http` crate, we're updating our client crates to use an intermediate type.

## Description

Update _client_ code to use `HttpRequest`. Some internal code still uses `http::Request` but those usages are limited.

## Testing
CI

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/rust-runtime/aws-inlineable/src/presigning.rs [hunk 1]
```diff
@@ -168,62 +170,46 @@ impl PresigningConfigBuilder {
...
+    pub fn uri(&self) -> &str {
         self.0.uri()
     }
 
     /// Returns any HTTP headers that need to go along with the request, except for `Host`,
     /// which should be sent based on the endpoint in the URI by the HTTP client rather than
     /// added directly.
-    pub fn headers(&self) -> &http::HeaderMap<http::HeaderValue> {
-        self.0.headers()
+    pub fn headers(&self) -> impl Iterator<Item = (&str, &str)> {
```
Comment (jdisanti) [inline aws/rust-runtime/aws-inlineable/src/presigning.rs:197]: Do we need the sensitive flag on headers for presigning?

File: aws/sdk/integration-tests/s3/tests/endpoints.rs [hunk 2]
```diff
@@ -3,6 +3,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#![cfg(feature = "test-util")]
```
Comment (jdisanti) [inline aws/sdk/integration-tests/s3/tests/endpoints.rs:6]: Now that integration tests are predicated on features, we should update the `check-aws-sdk-standalone-integration-tests` script here:
```
pushd "${tmp_dir}/aws/sdk/integration-tests"
cargo check --tests
popd
```
It should check with `--all-features`.

File: codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/util/Strings.kt [hunk 3]
```diff
@@ -18,10 +18,10 @@ fun String.doubleQuote(): String =
  */
 fun String.dq(): String = this.doubleQuote()
 
+val completeWords: List<String> = listOf("ipv4", "ipv6", "sigv4", "mib", "gib", "kib", "ttl")
```
Comment (jdisanti) [inline codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/util/Strings.kt:21]:
```suggestion
private val completeWords: List<String> = listOf("ipv4", "ipv6", "sigv4", "mib", "gib", "kib", "ttl")
```

File: rust-runtime/aws-smithy-checksums/src/http.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-checksums/src/http.rs:1]: I wonder why GitHub thinks this file is binary 🤔

File: aws/rust-runtime/aws-inlineable/src/presigning.rs [hunk 4]
```diff
@@ -168,62 +170,46 @@ impl PresigningConfigBuilder {
 ///
 /// **This struct has conversion convenience functions:**
 ///
-/// - [`PresignedRequest::to_http_request<B>`][Self::to_http_request] returns an [`http::Request<B>`](https://docs.rs/http/0.2.6/http/request/struct.Request.html)
+/// - [`PresignedRequest::to_http_02x_request<B>`][Self::to_http_03x_request] returns an [`http::Request<B>`](https://docs.rs/http/0.2.6/http/request/struct.Request.html)
```
Comment (jdisanti) [inline aws/rust-runtime/aws-inlineable/src/presigning.rs:173]:
```suggestion
/// - [`PresignedRequest::to_http_02x_request<B>`][Self::to_http_02x_request] returns an [`http::Request<B>`](https://docs.rs/http/0.2.6/http/request/struct.Request.html)
```

File: aws/rust-runtime/aws-inlineable/src/presigning.rs [hunk 5]
```diff
@@ -168,62 +170,46 @@ impl PresigningConfigBuilder {
...
+    pub fn headers(&self) -> impl Iterator<Item = (&str, &str)> {
+        self.0.headers().iter()
     }
 
     /// Given a body, convert this `PresignedRequest` into an `http::Request`
-    pub fn to_http_request<B>(self, body: B) -> Result<http::Request<B>, http::Error> {
-        let builder: http::request::Builder = self.into();
-
-        builder.body(body)
+    pub fn to_http_03x_request<B>(self, body: B) -> Result<http::Request<B>, BoxError> {
```
Comment (jdisanti) [inline aws/rust-runtime/aws-inlineable/src/presigning.rs:202]:
```suggestion
    pub fn to_http_02x_request<B>(self, body: B) -> Result<http::Request<B>, BoxError> {
```

File: rust-runtime/aws-smithy-runtime/src/client/http/test_util/dvr.rs [hunk 6]
```diff
@@ -106,11 +109,24 @@ impl<'a, B> From<&'a http::Request<B>> for Request {
...
     let mut out: HashMap<_, Vec<_>> = HashMap::new();
     for (header_name, header_value) in headers.iter() {
         let entry = out.entry(header_name.to_string()).or_default();
-        entry.push(header_value.to_str().unwrap().to_string());
+        entry.push(header_value.to_string());
+    }
+    out
+}
+
+fn headers_to_map_03x(headers: &HeaderMap) -> HashMap<String, Vec<String>> {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/test_util/dvr.rs:121]:
```suggestion
fn headers_to_map_02x(headers: &HeaderMap) -> HashMap<String, Vec<String>> {
```

File: rust-runtime/aws-smithy-runtime/src/client/http/test_util/dvr.rs [hunk 7]
```diff
@@ -119,7 +135,7 @@ impl<'a, B> From<&'a http::Response<B>> for Response {
     fn from(resp: &'a http::Response<B>) -> Self {
         let status = resp.status().as_u16();
         let version = format!("{:?}", resp.version());
-        let headers = headers_to_map(resp.headers());
+        let headers = headers_to_map_03x(resp.headers());
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/test_util/dvr.rs:138]:
```suggestion
        let headers = headers_to_map_02x(resp.headers());
```

File: rust-runtime/aws-smithy-runtime/src/client/http/test_util/replay.rs [hunk 8]
```diff
@@ -29,8 +29,11 @@ pub struct ReplayEvent {
 
 impl ReplayEvent {
     /// Creates a new `ReplayEvent`.
-    pub fn new(request: HttpRequest, response: HttpResponse) -> Self {
-        Self { request, response }
+    pub fn new(request: impl TryInto<HttpRequest>, response: impl TryInto<HttpResponse>) -> Self {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/test_util/replay.rs:32]: Nice!

File: rust-runtime/aws-smithy-runtime/src/client/orchestrator/endpoints.rs [hunk 9]
```diff
@@ -144,17 +143,34 @@ fn apply_endpoint(
     endpoint: &Endpoint,
     endpoint_prefix: Option<&EndpointPrefix>,
 ) -> Result<(), BoxError> {
-    let uri: Uri = endpoint.url().parse().map_err(|err| {
-        ResolveEndpointError::from_source("endpoint did not have a valid uri", err)
-    })?;
-
-    apply_endpoint_to_request_uri(request.uri_mut(), &uri, endpoint_prefix).map_err(|err| {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/orchestrator/endpoints.rs:151]: Is this old apply_endpoint method used anymore, or can it be removed?

**Review by jdisanti — COMMENTED** [review]
Great work on this! I only have minor comments so far.

File: rust-runtime/aws-smithy-runtime/src/client/orchestrator/endpoints.rs (hunk 9 above)
Comment (rcoh) [reply rust-runtime/aws-smithy-runtime/src/client/orchestrator/endpoints.rs:151]: there is one usage in the ECS provider. I think that can eventually use the endpoint resolver from the orchestrator though.

File: aws/rust-runtime/aws-inlineable/src/presigning.rs (hunk 1 above)
Comment (rcoh) [reply aws/rust-runtime/aws-inlineable/src/presigning.rs:197]: my take is this header is probably not _that_ sensitive to make it worthwhile to return a custom type here. Open to thoughts though.

**Review by jdisanti — APPROVED** [review]

**Review by ysaito1001 — APPROVED** [review]
Lots of places to update. Thanks for the work 🙇

File: rust-runtime/aws-smithy-checksums/src/http.rs
Comment (dtolnay) [reply rust-runtime/aws-smithy-checksums/src/http.rs:1]: Please see https://github.com/smithy-lang/smithy-rs/pull/3778. The original author put non-printable characters into this file.

**Pull request merged**